### DISCUSSION TITLE
error: value-initialization of reference type ‘nlohmann::basic_json<>&’

### DISCUSSION THREAD

**Discussion by fixbug666** [nlohmann/json#2888] [Q&A] [answered]
hi friends,  I use json in the code like this:
```cpp
using MsgHandler = std::function<void(const TcpConnectionPtr &conn, json &js, Timestamp)>;
class ChatService
{
public:
    void login(const TcpConnectionPtr &conn, json &js, Timestamp time);
private:
    map<int, MsgHandler> _msgHandlerMap;
};

_msgHandlerMap.emplace(LOGIN_MSG, std::bind(&ChatService::login, this, _1, _2, _3));
```

but in centos 7, gcc version is 4.8.5,  when i compile the code , there are always have some errors like this

```
/usr/include/c++/4.8.2/tuple: In instantiation of ‘constexpr std::_Head_base<_Idx, _Head, false>::_Head_base() [with long unsigned int _Idx = 1ul; _Head = nlohmann::basic_json<>&]’:
/usr/include/c++/4.8.2/type_traits:706:28:   required from ‘struct std::__is_default_constructible_impl<std::_Head_base<1ul, nlohmann::basic_json<>&, false> >’
.........
/usr/include/c++/4.8.2/tuple:132:22: error: value-initialization of reference type ‘nlohmann::basic_json<>&’
       : _M_head_impl() { }
```

but it complies ok in higher gcc version and gcc 4.7 version,  what is the problem?

**Comment by nlohmann** [comment, marked as answer]
We do not support GCC 4.7, and GCC 4.8 still has some bugs in the C++11 support. Your error message looks as if you are using GCC 4.8.5, but the headers of 4.8.2 - can you double check this is in order?

**Comment by wangwenqiangGitHub** [comment]
the json lib version 3.10 still has this isue:
arm-linux-gnueabi/include/c++/4.9.4/tuple:142:42: error: invalid initialization of non-const reference of type 'nlohmann::basic_json<>&' from an rvalue of type 'std::basic_string'
: _M_head_impl(std::forward<_UHead>(__h)) { }

**Reply by nlohmann** [reply]
Can you provide a small code example?

**Reply by falbrechtskirchinger** [reply]
See #2147. Starts compiling from GCC 5.2 onward (tested this earlier with the code from #2147).

**Comment by haoden1313** [comment]
请问解决了吗
